pkg/controller/disruption/disruption.go

```diff
@@ -777,9 +780,21 @@ func (dc *DisruptionController) buildDisruptedPodMap(pods []*v1.Pod, pdb *policy
 // implement the  "fail open" part of the design since if we manage to update
 // this field correctly, we will prevent the /evict handler from approving an
 // eviction when it may be unsafe to do so.
-func (dc *DisruptionController) failSafe(pdb *policy.PodDisruptionBudget) error {
+func (dc *DisruptionController) failSafe(pdb *policy.PodDisruptionBudget, err error) error {
```

Comment:
Would it be possible to add another condition to indicate we're hitting failSafe rather than just disruptions allowed.  We could consume that info here: https://github.com/kubernetes/kubernetes/blob/release-1.20/pkg/registry/core/pod/storage/eviction.go#L208

Otherwise, we might evict pods with DesiredHealthy && CurrentHealthy being stale.

Maybe we can still achieve this by testing Reason == policy.SyncFailedReason if we have a util function, but sync failure seems like it should be it's own status condition rather than embedded in DisruptionsAllowed.  WDYT?

Reply:
We could add a separate condition for this, but I wanted to avoid creating more conditions than necessary. So if we did this, it would be something like a `SyncFailed` condition? I'm not sure what kind of reasons we would use here, but maybe we could enumerate the "likely" errors we might encounter.

And would conditions be the appropriate way to handle this? The `pdb.Status.DisruptionsAllowed` would be useful here since it is updated by the eviction api, but I guess it doesn't work here since we don't know if a value of zero means we have the minimum number of pods to satisfy the PDB or if we have fewer than we need. Would it be an option to also decrement the `pdb.Status.CurrentHealthy` count after eviction?

Reply:
> Otherwise, we might evict pods with DesiredHealthy && CurrentHealthy being stale.

is that true? failSafe sets DisruptionsAllowed to 0, which prevents evictions, right?

if SyncFailed=True always implies DisruptionAllowed=False, that seems like a good use of the distinct reason field on a DisruptionAllowed=False condition rather than a separate condition

Reply:
Currently, DisruptionsAllowed is an integer, and this is the 'budget' used to determine how many pods are evicted.  If we are hoping to explicitly disallow eviction because the PDB has gone stale, we need another field.

I don't have a strong preference on the field name or type, so long as it's for that specific purpose.

Reply:
the same controller that would add the "I am stale because I can't sync" condition is already setting DisruptionsAllowed to 0 when it can't sync... I don't see why another field is required

Reply:
Because eviction sometimes doesn't consider DisruptionsAllowed if a pod is unready and there is enough CurrentHealthy: https://github.com/kubernetes/kubernetes/blob/release-1.20/pkg/registry/core/pod/storage/eviction.go#L208

We need to know if CurrentHealthy is stale or not, we *want* to remove unready pods when DisruptionsAllowed==0 && CurrentHealthy >= DesiredHealthy.

Reply:
I see.

Reply:
Could we address this by decrementing the `CurrentHealthy` field at the same time we are decrementing `DisruptionsAllowed` in `EvictionREST#checkAndDecrement`? 
 https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L333 ? 

I think that would make sure the value of that field isn't stale (at least to the same extent as the `DisruptionsAllowed` field). And it seems reasonable to decrement this value at that point since we are just about to terminate a healthy pod.

Reply:
That seems reasonable.  I don't have a strong preference on what the signal is, just as long as it is able to accomplish what we need.  If we go this route, please let's add some comments so we don't regress.  This would also prevent us from needing specific tests in the eviction API, but we will ideally have tests on the PDB side.

Reply:
Looking at this again, my proposal in the previous comment doesn't actually address the problem. To handle the issue where sync fails, we would either need to set `CurrentHealthy` to 0 at the same time we are setting `DisruptionsAllowed` to 0 in the [failSafe function](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption.go#L782), introduce a new field, or let the eviction API look at the conditions. 

Forcing the value of `CurrentHealthy` to 0 doesn't seem great as someone might use this field (although the value will be stale in this situation). 

Introducing a new field seems a bit superfluous since the information is already available in the condition.

I think I would prefer that we handle this with the suggested `DisruptionsAllowed` condition, and we just let the eviction API look at the `Reason` field. Conditions are part of the public API, so I don't think there is any major issues with doing it this way.

@liggitt Do you have a preference here? It seems like this is something that could block the PDB to GA effort unless we find a solution pretty quickly.

@michaelgugino Is there an existing github issue describing this problem? If not, can you create one so we can move the discussion there? It is mostly unrelated to this PR.

Reply:
> Because eviction sometimes doesn't consider DisruptionsAllowed if a pod is unready and there is enough CurrentHealthy: https://github.com/kubernetes/kubernetes/blob/release-1.20/pkg/registry/core/pod/storage/eviction.go#L208
> 
> We need to know if CurrentHealthy is stale or not, we _want_ to remove unready pods when DisruptionsAllowed==0 && CurrentHealthy >= DesiredHealthy.

actually, I'm not sure I understand why we need to check CurrentHealthy / DesiredHealthy in this case... removing a not-ready pod should always be ok, right?

Reply:
If we do end up needing to determine if the controller is failing to sync, checking for `DisruptionsAllowed=False, Reason= SyncFailed` could be reasonable, *but* observers cannot depend on this condition being added/removed reliably until clusters are completely upgraded to v1.21. That means code we release in 1.21 has to tolerate the condition not being added/removed reliably:

* a multi-server cluster where the active controller-manager is still 1.20 will not have this condition added
* a multi-server cluster where the leader controller-manager changed to 1.21, added the SyncFailed condition, then changed back to 1.20 will not have this condition removed

Reply:
I added the discussion issue here: https://github.com/kubernetes/kubernetes/issues/99598

The current patch is fine if we don't like manipulating CurrentHealthy.  My suggestion is to add a util function to detect this specific status, and ensure that document the intent to others that 'SyncFailed' means you probably should not rely on the PDB status counts as they may be stale.

This is somewhat of an edge case that probably we should have accounted for prior to the eviction API change, but I missed it.

Reply:
Replied in https://github.com/kubernetes/kubernetes/issues/99598

Reply:
addressing this in a separate PR sounds good to me. I don't think this change makes the issue worse, and gives us a potential tool to address the issue in a follow up